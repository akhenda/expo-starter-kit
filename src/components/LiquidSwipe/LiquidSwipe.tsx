import { useState } from 'react';

import { images } from '@src/theme';

import { Slide, SlideProps } from './Slide';
import { Slider } from './Slider';

const slides: SlideProps['slide'][] = [
  {
    color: '#F2A1AD',
    description: 'Hot or cold, our dessert recipes can turn an average meal into a memorable event',
    picture: images.onboarding.liquidSwipe[0],
    title: 'Dessert Recipes',
  },
  {
    color: '#0090D6',
    description: 'Discover healthy recipes that are easy to do with detailed cooking instructions from top chefs',
    picture: images.onboarding.liquidSwipe[4],
    title: 'Healthy Foods',
  },
  {
    color: '#69C743',
    description: 'explore recipes by food type, preparation method, cuisine, country and more',
    picture: images.onboarding.liquidSwipe[3],
    title: 'Easy Meal Ideas',
  },
  {
    color: '#FB3A4D',
    description: 'Browse thousands of curated recipes from top chefs, each with detailled cooking instructions',
    picture: images.onboarding.liquidSwipe[1],
    title: '10000+ Recipes',
  },
  {
    color: '#F2AD62',
    description: 'Browse our best themed recipes, cooking tips, and how-to food video & photos',
    picture: images.onboarding.liquidSwipe[2],
    title: 'Video Tutorials',
  },
];

export const assets = slides.map(({ picture }) => picture);

const LiquidSwipe = () => {
  const [index, setIndex] = useState(0);

  const prev = slides[index - 1];
  const next = slides[index + 1];

  return (
    <Slider
      key={index}
      index={index}
      setIndex={setIndex}
      prev={prev && <Slide slide={prev} />}
      next={next && <Slide slide={next} />}
    >
      <Slide slide={slides[index] as SlideProps['slide']} />
    </Slider>
  );
};

export default LiquidSwipe;
