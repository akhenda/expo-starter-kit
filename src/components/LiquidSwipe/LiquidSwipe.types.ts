import type { StyleProp, ViewProps, ViewStyle } from 'react-native';

export interface LiquidSwipeProps extends ViewProps {
  /**
   * Extra styles for the container
   */
  style?: StyleProp<ViewStyle>;
}
