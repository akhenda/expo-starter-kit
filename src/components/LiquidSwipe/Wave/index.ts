export * from './Wave';
export { default as Wave } from './Wave';
export * from './Wave.types';
