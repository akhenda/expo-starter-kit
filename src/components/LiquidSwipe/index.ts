export { default as LiquidSwipe } from './LiquidSwipe';
export * from './LiquidSwipe.types';
