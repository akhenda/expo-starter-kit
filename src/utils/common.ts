import { Linking } from 'react-native';

/**
 * A "modern" sleep statement.
 *
 * @param ms The number of milliseconds to wait.
 */
export const delay = (ms: number) => {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
};

/**
 * Helper for opening a give URL in an external browser.
 */
export function openLinkInBrowser(url: string) {
  Linking.canOpenURL(url).then((canOpen) => canOpen && Linking.openURL(url));
}
