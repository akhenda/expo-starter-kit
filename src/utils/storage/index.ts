export { asyncStorage } from './async-storage';
export { mmkvStorage } from './mmkv';
export { default as mmkv } from './mmkv';
