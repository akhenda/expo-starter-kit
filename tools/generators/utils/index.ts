export * from './getPath';
export * from './prettierTransform';
export * from './requireField';
