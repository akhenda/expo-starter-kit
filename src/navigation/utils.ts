import React, { useCallback, useEffect, useLayoutEffect, useRef, useState } from 'react';
import { BackHandler } from 'react-native';
import { createNavigationContainerRef, NavigationState, useNavigation } from '@react-navigation/native';

import type { PersistNavigationConfig } from '@config/config.base';
import Config from '@src/config';
import { useIsMounted } from '@src/hooks/useLifecycle';
import { mmkvStorage as storage } from '@src/services/storage';

import type { RootStackParamList } from './navigators/RootNavigator';

export const NAVIGATION_PERSISTENCE_KEY = '@NAVIGATION_STATE';

export const navigationRef = createNavigationContainerRef<RootStackParamList>();

/**
 * Gets the current screen from any navigation state.
 */
export function getActiveRouteName(state?: NavigationState<RootStackParamList>): keyof RootStackParamList {
  if (!state || !state.index) return 'App Stack';

  const route = state.routes[state.index];

  if (!route) return 'App Stack';

  // Found the active route -- return the name
  if (!route.state) return route.name;

  // Recursive call to deal with nested routers
  return getActiveRouteName(route.state as NavigationState<RootStackParamList>);
}

/**
 * Hook that handles Android back button presses and forwards those on to
 * the navigation or allows exiting the app.
 */
export function useBackButtonHandler(canExit: (routeName: keyof RootStackParamList) => boolean) {
  // The reason we're using a ref here is because we need to be able
  // to update the canExit function without re-setting up all the listeners
  const canExitRef = useRef(canExit);

  useEffect(() => {
    canExitRef.current = canExit;
  }, [canExit]);

  useEffect(() => {
    // We'll fire this when the back button is pressed on Android.
    const onBackPress = () => {
      if (!navigationRef.isReady()) return false;

      // grab the current route
      const routeName = getActiveRouteName(navigationRef.getRootState() as NavigationState<RootStackParamList>);

      // are we allowed to exit?
      if (canExitRef.current(routeName)) {
        // exit and let the system know we've handled the event
        BackHandler.exitApp();

        return true;
      }

      // we can't exit, so let's turn this into a back action
      if (navigationRef.canGoBack()) {
        navigationRef.goBack();

        return true;
      }

      return false;
    };

    // Subscribe when we come to life
    BackHandler.addEventListener('hardwareBackPress', onBackPress);

    // Unsubscribe when we're done
    return () => BackHandler.removeEventListener('hardwareBackPress', onBackPress);
  }, []);
}

/**
 * This helper function will determine whether we should enable navigation persistence
 * based on a config setting and the __DEV__ environment (dev or prod).
 */
function navigationRestoredDefaultState(persistNavigation: PersistNavigationConfig) {
  if (persistNavigation === 'always') return false;
  if (persistNavigation === 'dev' && __DEV__) return false;
  if (persistNavigation === 'prod' && !__DEV__) return false;

  // all other cases, disable restoration by returning true
  return true;
}

/**
 * Custom hook for persisting navigation state.
 */
export function useNavigationPersistence(persistenceKey: string = NAVIGATION_PERSISTENCE_KEY) {
  const isMounted = useIsMounted();
  const routeNameRef = useRef<string | undefined>();
  const initNavState = navigationRestoredDefaultState(Config.persistNavigation);

  const [initialNavigationState, setInitialNavigationState] = useState<NavigationState<RootStackParamList>>();
  const [isRestored, setIsRestored] = useState(initNavState);

  const onNavigationStateChange = (state?: NavigationState) => {
    const previousRouteName = routeNameRef.current;
    const currentRouteName = getActiveRouteName(state as NavigationState<RootStackParamList>);

    if (previousRouteName !== currentRouteName) {
      // track screens.
      if (__DEV__)
        // eslint-disable-next-line no-console
        console.tron.display?.({
          name: 'currentRouteName',
          preview: currentRouteName,
          value: currentRouteName,
        });
    }

    // Save the current route name for later comparison
    routeNameRef.current = currentRouteName;

    // Persist state to storage
    storage.save(persistenceKey, state);
  };

  const restoreState = useCallback(async () => {
    try {
      const state = (await storage.load(persistenceKey)) as NavigationState<RootStackParamList>;

      if (state) setInitialNavigationState(state);
    } finally {
      if (isMounted()) setIsRestored(true);
    }
  }, [persistenceKey, isMounted]);

  useEffect(() => {
    if (!isRestored) restoreState();
  }, [isRestored, restoreState]);

  return { initialNavigationState, isRestored, onNavigationStateChange, restoreState };
}

/**
 * A hook that can be used to easily set the Header of a react-navigation screen from within the screen's component.
 *
 * - [Documentation and Examples](https://github.com/infinitered/ignite/blob/master/docs/Utils-useHeader.md)
 */
export function useHeader(header: React.ReactNode, deps: unknown[] = []) {
  const navigation = useNavigation();

  useLayoutEffect(() => {
    navigation.setOptions({
      header,
      headerShown: true,
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [...deps, header, navigation]);
}

/**
 * use this to navigate without the navigation
 * prop. If you have access to the navigation prop, do not use this.
 * More info: https://reactnavigation.org/docs/navigating-without-navigation-prop/
 */
export function navigate(...args: Parameters<typeof navigationRef.navigate>): void {
  if (navigationRef.isReady()) navigationRef.navigate(...args);
}

export function goBack() {
  if (navigationRef.isReady() && navigationRef.canGoBack()) navigationRef.goBack();
}

export function resetRoot(params = { index: 0, routes: [] }) {
  if (navigationRef.isReady()) navigationRef.resetRoot(params);
}
