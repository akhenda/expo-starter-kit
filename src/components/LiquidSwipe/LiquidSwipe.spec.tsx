import { cleanup, screen } from '@testing-library/react-native';

import { render } from '@src/utils/tests';

import LiquidSwipe from './LiquidSwipe';

afterEach(cleanup);

it('renders correctly', async () => {
  render(<LiquidSwipe />);

  expect(screen.toJSON()).toMatchSnapshot();
});
