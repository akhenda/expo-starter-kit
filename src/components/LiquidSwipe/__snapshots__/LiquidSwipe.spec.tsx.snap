// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  collapsable={false}
  handlerTag={1}
  handlerType="PanGestureHandler"
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  style={
    {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <RNSVGSvgView
    bbHeight="100%"
    bbWidth="100%"
    focusable={false}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "flex": 0,
          "height": "100%",
          "width": "100%",
        },
      ]
    }
  >
    <RNSVGGroup>
      <RNSVGDefs>
        <RNSVGRadialGradient
          cx="50%"
          cy="35%"
          fx="50%"
          fy="35%"
          gradient={
            [
              0,
              -1,
              1,
              -876115,
            ]
          }
          gradientTransform={null}
          gradientUnits={0}
          name="gradient"
          rx="50%"
          ry="50%"
        />
      </RNSVGDefs>
      <RNSVGRect
        fill={
          {
            "brushRef": "gradient",
            "type": 1,
          }
        }
        height="1334"
        propList={
          [
            "fill",
          ]
        }
        width="750"
        x="0"
        y="0"
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    style={
      {
        "alignItems": "center",
        "bottom": 0,
        "left": 0,
        "padding": 75,
        "paddingTop": 150,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <Image
      source={1}
      style={
        {
          "height": 675,
          "width": 675,
        }
      }
    />
    <View>
      <Text
        style={
          {
            "color": "white",
            "fontFamily": "SFProDisplay-Bold",
            "fontSize": 48,
            "marginBottom": 16,
            "textAlign": "center",
          }
        }
      >
        Dessert Recipes
      </Text>
      <Text
        style={
          {
            "color": "white",
            "fontFamily": "SFProDisplay-Regular",
            "fontSize": 18,
            "textAlign": "center",
          }
        }
      >
        Hot or cold, our dessert recipes can turn an average meal into a memorable event
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <RNCMaskedView
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        pointerEvents="none"
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <RNSVGSvgView
          bbHeight="100%"
          bbWidth="100%"
          focusable={false}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              [
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                {
                  "transform": [
                    {
                      "rotateY": "180deg",
                    },
                  ],
                },
              ],
              {
                "flex": 0,
                "height": "100%",
                "width": "100%",
              },
            ]
          }
        >
          <RNSVGGroup
            matrix={
              [
                1,
                0,
                0,
                1,
                0,
                0,
              ]
            }
          >
            <RNSVGPath
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              propList={
                [
                  "fill",
                ]
              }
            />
          </RNSVGGroup>
        </RNSVGSvgView>
      </View>
      <RNSVGSvgView
        bbHeight="100%"
        bbWidth="100%"
        focusable={false}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "flex": 0,
              "height": "100%",
              "width": "100%",
            },
          ]
        }
      >
        <RNSVGGroup>
          <RNSVGDefs>
            <RNSVGRadialGradient
              cx="50%"
              cy="35%"
              fx="50%"
              fy="35%"
              gradient={
                [
                  0,
                  -8202497,
                  1,
                  -16740138,
                ]
              }
              gradientTransform={null}
              gradientUnits={0}
              name="gradient"
              rx="50%"
              ry="50%"
            />
          </RNSVGDefs>
          <RNSVGRect
            fill={
              {
                "brushRef": "gradient",
                "type": 1,
              }
            }
            height="1334"
            propList={
              [
                "fill",
              ]
            }
            width="750"
            x="0"
            y="0"
          />
        </RNSVGGroup>
      </RNSVGSvgView>
      <View
        style={
          {
            "alignItems": "center",
            "bottom": 0,
            "left": 0,
            "padding": 75,
            "paddingTop": 150,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <Image
          source={1}
          style={
            {
              "height": 675,
              "width": 675,
            }
          }
        />
        <View>
          <Text
            style={
              {
                "color": "white",
                "fontFamily": "SFProDisplay-Bold",
                "fontSize": 48,
                "marginBottom": 16,
                "textAlign": "center",
              }
            }
          >
            Healthy Foods
          </Text>
          <Text
            style={
              {
                "color": "white",
                "fontFamily": "SFProDisplay-Regular",
                "fontSize": 18,
                "textAlign": "center",
              }
            }
          >
            Discover healthy recipes that are easy to do with detailed cooking instructions from top chefs
          </Text>
        </View>
      </View>
    </RNCMaskedView>
    <View
      style={
        {
          "alignItems": "center",
          "borderRadius": 25,
          "height": 50,
          "justifyContent": "center",
          "left": 750,
          "opacity": 1,
          "position": "absolute",
          "top": 642,
          "width": 50,
        }
      }
    >
      <Text />
    </View>
  </View>
</View>
`;
